### ISSUE TITLE
Callbacks with compound elements containing lifetimes

### ISSUE THREAD

**Issue by ratmice** [maciejhirsz/logos#118]
If we have a token like the following silly example:

```
enum Token<'a> {
  #[regex("[a-b]+", |lexer| {
    let slice = lexer.slice();
    (slice, slice[0..1] == "a")
   })]
  AB((&'a str, bool))
}
```

Alternately switching the enumeration member to:
```
   enum Foo<S> {
      A(S),
      B(S),
}

#[regex("[a-b]+", |lexer| {
  let slice = lexer.slice();
  if slice[0..1] == "a" { Foo::A(slice) }
  else { Foo::B(slice) }
})]
AB(Foo<&'a str>)
```
In both the above cases you tend to get something like:

> 12 | #[derive(Debug, Clone, Logos)]
>    |                        - help: consider introducing lifetime `'a` here: `'a,`
> 
> 57 |     AB((&'a str, bool)),
>    |               ^^ undeclared lifetime

with The 'here:' pointing to the Logos, which is somewhat odd.

Removing the lifetimes from the enum variant however *does* work:

```
  #[regex("[a-b]+", |lexer| {
    let slice = lexer.slice();
    (&slice[0..1] == "a", &slice[0..1] == "b")
   })]
  AB((bool, bool))
```

Are there any thoughts/tricks on how to construct variants with compound elements with member bounded by lifetimes?

**Comment by maciejhirsz** [comment]
Yeah, that's not supported yet. The trick is that I need to recursively run through the type declaration and swap lifetimes to make sure they match declarations in generated code, it's not hard to do.

Will give it a go tomorrow. I was also already thinking about unpacking tuples for variants with more than one field, but I'm not sure how easy or hard that would be yet.

**Comment by ratmice** [comment]
I did find a workaround which works for my use case:
Since I was planning on using the `bool` in `lex.filter(|..| ...)`,
I can just do the string comparison in the filter closure moving it out of the callback.

**Comment by maciejhirsz** [comment]
0.11.0-rc4 introduced a [skip](https://github.com/maciejhirsz/logos/blob/0e914ef4ce0043b38f74b24d602b49f2b519f579/tests/tests/simple.rs#L38) mechanism, I was already planning to expand that to an enum like:

```
enum Maybe<Token> {
    Emit(Token),
    Skip,
}
```

So you could effectively do a filter in the callback. Also something for tomorrow :).

**Comment by maciejhirsz** [comment]
Released 0.11.0-rc5 with features mentioned above, named the enum `Filter` instead of `Maybe` to better reflect its function.
